### PR TITLE
Better method call error messages

### PR DESCRIPTION
**Pull request by jackh726**
Rebase/continuation of #71827

~Based on #92360~
~Based on #93118~

There's a decent description in #71827 that I won't copy here (for now at least)

In addition to rebasing, I've tried to restore most of the original suggestions for invalid arguments. Unfortunately, this does make some of the errors a bit verbose. To fix this will require a bit of refactoring to some of the generalized error suggestion functions, and I just don't have the time to go into it right now.

I think this is in a state that the error messages are overall better than before without a reduction in the suggestions given.

~I've tried to split out some of the easier and self-contained changes into separate commits (mostly in #92360, but also one here). There might be more than can be done here, but again just lacking time.~

r? @estebank as the original reviewer of #71827

### CODE REVIEW

**Comment by jackh726** [comment]
I will resolve the build failures after first round of review or after #92360 gets merged

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:428]: This could be its own change, as it isn't reliant on the subsequent checks.

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:644]: Everything since line 430 could be its own struct and impl, so that we can remove all of this logic out of the diagnostic code and make it clearer what is happening to anyone reading the code as prose.

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:656]:
```suggestion
            //   fn some_func(_a: i32, _b: i32) {}
```

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:681]: we have some code to suggest default values for *some* types, that could be used here.

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:687]: We might want to create a specific `ObligationCauseCode` for this case, IIF there's value in that.

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:727]: Also check that `ty` isn't `_` or `Err`.

Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:727]:
```suggestion
                        let arg_type = if let Some((_, ty)) = final_arg_types[arg_idx] {
                            format!(" of type `{}`", ty)
```

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:737]: Let's think about an alternative to "did you mean"

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:649]: Add comment explaining that if multiple changes are replaced with "did you mean"

File: src/test/ui/argument-suggestions/basic.stderr
Comment (estebank) [inline src/test/ui/argument-suggestions/basic.stderr:16]: I think that this case is common enough that we might want to special case it slightly. What do you think of

```
error[E0308]: mismatched types: expected `u32`, found floating-point number
  --> $DIR/basic.rs:20:5
   |
LL |     invalid(1.0);
   |     ------- ^^^ expected `u32`, found floating-point number
   |     |
   |     arguments to this function are incorrect
   |
note: function defined here
  --> $DIR/basic.rs:13:4
   |
LL | fn invalid(_i: u32) {}
   |    ^^^^^^^
```

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:54]: Maybe we should move these to `arg_matrix.rs` too.

File: src/test/ui/argument-suggestions/basic.stderr
Comment (estebank) [inline src/test/ui/argument-suggestions/basic.stderr:22]: This should be

```suggestion
   |     ----- ^^ unexpected argument
```

File: src/test/ui/argument-suggestions/basic.stderr
Comment (estebank) [inline src/test/ui/argument-suggestions/basic.stderr:32]: Ideally, given the conversation we had yesterday, I would like this to be, but it doesn't have to be in this PR

```suggestion
help: remove the extra argument
   |
LL -     extra("");
LL +     extra();
   |
```

File: src/test/ui/argument-suggestions/basic.stderr
Comment (jackh726) [reply src/test/ui/argument-suggestions/basic.stderr]: The output has changed a bit (the "expected" note is now a label again). I'm on the fence here: I do think we can have a little bit more specific error message when there is only one problem. However, getting this to be "just right" is probably pretty subjective and I don't really want to put in a ton of time on this or have this sit.

File: src/test/ui/argument-suggestions/basic.stderr
Comment (jackh726) [reply src/test/ui/argument-suggestions/basic.stderr:17]: Is there any way to do this other than changing the primary span of the error and the span of the label? (Though, then the message would be different...)

File: src/test/ui/argument-suggestions/basic.stderr
Comment (jackh726) [reply src/test/ui/argument-suggestions/basic.stderr:27]: Let's leave this for followup.

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (jackh726) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:577]: Can we leave this as a followup?

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (jackh726) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs]: Can we leave this for followup, or do you think it needs to be done in this PR?

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (jackh726) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs]: In #93118, I make error emission a bit more lazy and more like in this PR. It's not a 100% transplant, but close enough the diff should be nicer.

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (estebank) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs]: Follow up PR is fine.

File: src/test/ui/typeck/struct-enum-wrong-args.stderr
Comment (estebank) [inline src/test/ui/typeck/struct-enum-wrong-args.stderr:23]: These changes look great!

I think we could potentially change the wording in the future so we don't talk about arguments and instead talk about fields, but that's independent of this PR.

File: src/test/ui/typeck/issue-89856.stderr
Comment (estebank) [inline src/test/ui/typeck/issue-89856.stderr:15]: This makes me think: could we special case this to point at the `fn` args here? Potential future change.

File: src/test/ui/type/type-ascription-instead-of-initializer.stderr
Comment (estebank) [inline src/test/ui/type/type-ascription-instead-of-initializer.stderr:19]: We are losing the span pointing at the associated fn here. Note to self: check where that regression might be happening.

File: src/test/ui/span/missing-unit-argument.stderr
Comment (estebank) [inline src/test/ui/span/missing-unit-argument.stderr:22]: Should we special case this to provide a single `span_label` saying "two arguments of type `()` and `()` are missing"?

File: src/test/ui/type/type-ascription-instead-of-initializer.stderr
Comment (jackh726) [reply src/test/ui/type/type-ascription-instead-of-initializer.stderr:19]: That would be because somehow we changed from `def_ident_span` to `get_if_local`. Fixing...

File: src/test/ui/span/missing-unit-argument.stderr
Comment (jackh726) [reply src/test/ui/span/missing-unit-argument.stderr:22]: Good idea. Might be worth thinking about consecutive missing args in general:
```
an argument of type `X` is missing
``` 
```
two arguments of type `X` and `Y` are missing
```
```
three arguments of type `X`, `Y`, and `Z` are missing
```
```
multiple missing arguments
```
(the last one is still useful because of the "function defined here" note

I think we only need to fall back here for *consecutive* arguments; if missing args are interspersed by others, the spans will be different and it won't be confusing

We also can probably have a better span here: ideally probably an "empty" span between the parantheses.

Comment (estebank) [reply src/test/ui/span/missing-unit-argument.stderr:22]: What I like about the current output is that it is simple and effective regardless of count: "expected 2 arguments"/"supplied 0 arguments" doesn't need that much logic to work well in all cases.

An empty span between the parentheses will show up in a problematic way, because it would display as an underline under the closing `)`, and we wouldn't want that. But it can point at the `()`, like it does now.

Comment (jackh726) [reply src/test/ui/span/missing-unit-argument.stderr:22]: I'll play around a bit with this. I though you could point at an "empty" span, with just like a line.

**Comment by jackh726** [comment]
Okay @estebank, I think I'm caught up with reviews that need to be done in this PR. This is ready for another round of review. (I'm going to squash the commits before merge, but leaving them for now for better review)

A brief list of reviews made but not changed in this PR:
- Suggest default values instead of e.g. `{i32}`
- Create an ObligationCauseCode instead of misc (don't think this is needed right now)
- Change "did you mean" to something else
  - I did make this get emitted less; now if there are multiple errors of the same type, then that just get pluralized
- "diff" view of removal/addition in error
- Talk about fields instead of arguments for tuple structs/enums
- Pointing at fn args in "function defined here"

A brief list of "big" changes made since reviews:
- ArgMatrix is now in a separate file with it's logic basically completely separate
- We don't emit "of type" if type is error or has inference vars
- If there is a single error, we emit that as we did before
- We merge adjacent missing argument errors into a single one

**Comment by Quantumplation** [comment]
Hey @jackh726, thanks for taking this and helping update it, really really appreciate that! I've been busy launching a startup, and that's been *super* crazy, so I hadn't had time to revisit.

I'd still like to help out, if there are some small things that I could help with!

**Comment by jackh726** [comment]
At this point, I think this waiting for another round of review from @estebank.

There are some followup items above that can be done "in parallel" as a separate PR based on this - and just rebased when this lands.

File: src/test/ui/span/coerce-suggestions.stderr
Comment (estebank) [inline src/test/ui/span/coerce-suggestions.stderr:25]: Note to self: we should suggest `&mut y` here.

File: src/test/ui/span/E0057.stderr
Comment (estebank) [inline src/test/ui/span/E0057.stderr:12]: Somewhat surprised we don't point at the definition of `f` here.

File: src/test/ui/resolve/resolve-primitive-fallback.stderr
Comment (estebank) [inline src/test/ui/resolve/resolve-primitive-fallback.stderr:32]: Future work: special case support for noticing that a function argument was supposed to be a type param.

File: src/test/ui/methods/method-call-err-msg.stderr
Comment (estebank) [inline src/test/ui/methods/method-call-err-msg.stderr:47]: **Please fix** there's a bug in the span here, it should be replacing the `(0)` as well, and it isn't.

File: src/test/ui/methods/method-call-err-msg.stderr
Comment (estebank) [inline src/test/ui/methods/method-call-err-msg.stderr:31]: Same here. I'm guessing there's a small issue with the span used for method calls.

File: src/test/ui/indexing-requires-a-uint.stderr
Comment (estebank) [inline src/test/ui/indexing-requires-a-uint.stderr:23]: We lost this suggestion here? Could you check if we can get it back? Not a dealbreaker.

File: src/test/ui/hrtb/issue-58451.stderr
Comment (estebank) [inline src/test/ui/hrtb/issue-58451.stderr:17]: I'm worried that the `{_}` placeholder might be a bit abstract, but we can iterate on this after landing the PR.

File: src/test/ui/fn/fn-item-type.rs
Comment (estebank) [inline src/test/ui/fn/fn-item-type.rs:48]: Slightly surprised at this change.

**Review by estebank — APPROVED** [review]
I'm ok with the current changes, modulo the invalid suggestions. There's some outstanding work to bring this to par, but lets land this asap.

r=me after fixing the span for incorrect method args suggestion.

**Comment by jackh726** [comment]
I've rebased and fixed the suggestions. Rebasing over the tuple suggestion changes in #91530 was a bit difficult, and the code is a bit duplicated. But I'd rather land this now and cleanup in-tree, rather than risk bitrotting even more.

@bors r=estebank p=1 rollup=never

**Comment by pnkfelix** [comment]
Visiting for weekly performance triage.

* The main regression here was to externs but there's a lot of historical noise in the data for externs, and not trustworthy.

@rustbot label: -perf-regression-triaged

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
Comment (flip1995) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:432]: @jackh726 This line causes an ICE when compiling `chalk` with `cargo check --all-targets --all-features`. AFAICT the issue here is that `input_idx` is in the range of `0..expected_input_tys.len()`, but the amount of passed types could be smaller. I don't have a reproducer yet. 

I don't think returning the `input_idx` based on the `final_arg_types.len()` would fix the issue, but rather reverse the issue with `expected_input_tys`, if more arguments are passed than expected.

Maybe this `Error` kind should hold both indices :thinking:

Comment (flip1995) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:432]: Minimal reproducer:

```rust
fn f(_: usize, _: &usize, _: usize) {}

fn arg<T>() -> T { todo!() }

fn main() {
    let x = arg(); // `x` must be inferred
    f(&x, ""); // The reference on `&x` is important to reproduce the ICE
}
```

Comment (flip1995) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:432]: Opened https://github.com/rust-lang/rust/issues/96638

**Pull request merged**
